rs/registry/canister/chunks/src/lib.rs

```diff
@@ -0,0 +1,58 @@
+use ic_crypto_sha2::Sha256;
+use ic_stable_structures::{Memory, StableBTreeMap};
+
+const CHUNK_SIZE_BYTES: usize = 1_800_000;
+
+/// Splits "monolithic" blobs into chunks, and stores the chunks. Chunks are
+/// keyed by their SHA256.
+///
+/// Deleting is not supported (because there seems to be no need, but could be
+/// added later).
+///
+/// Since chunks are keyed by their SHA256, modifying elements does not make sense.
+pub struct Chunks<MyMemory: Memory> {
+    chunk_sha256_to_content: StableBTreeMap<Vec<u8>, Vec<u8>, MyMemory>,
+}
+
+impl<MyMemory: Memory> Chunks<MyMemory> {
+    /// Like StableBTreeMap::init.
+    pub fn init(memory: MyMemory) -> Self {
+        Self {
+            chunk_sha256_to_content: StableBTreeMap::init(memory),
+        }
+    }
+
+    /// Returns the SHA256s of the chunks, which are the keys of the chunks.
+    ///
+    /// The chunks can be fetched via the `get` method. (Ofc, when you
+    /// concatenate the chunks, you get back the original monolithic blob.)
+    pub fn upsert_monolithic_blob(&mut self, mut monolithic_blob: Vec<u8>) -> Vec<Vec<u8>> {
+        let mut result = vec![];
+
+        // Split monolithic blob into chunks, and insert them into self.chunk_sha256_to_content.
+        while !monolithic_blob.is_empty() {
+            let next_chunk_size = std::cmp::min(CHUNK_SIZE_BYTES, monolithic_blob.len());
+            let chunk = monolithic_blob
+                .drain(..next_chunk_size)
+                .collect::<Vec<u8>>();
+            let sha256 = Sha256::hash(&chunk).to_vec();
+
+            let _old_content = self.chunk_sha256_to_content.insert(sha256.clone(), chunk);
```

Comment:
Is it faster to check if the key exists than to re-insert the data?  If the key exists, you know the content is the same, so you could save some instructions and not rewrite it.  I'm not really sure how impactful that actually would be, but it costs money to create dirty pages in memory.

Reply:
Maybe. I think most monolithic_blobs will not be upserted more than once.

We can definitely optimize this later if we find that in practice, it would give us a big win.

I think it is better to start simple.

> Premature optimization is the root of all evil.
> --Donald Knuth